File: tests/07-negative_authorization.ts
import * as anchor from "@anchor-lang/core";
import { Program } from "@anchor-lang/core";
import { Keypair, TransactionInstruction } from "@solana/web3.js";
import assert from "assert";

import { R3SessionKeys } from "../target/types/r3_session_keys";
import * as sessionKeys from "./utils/pda";
import {
  ensureProgramInitialized,
  executeWithSession,
  sendExpectError,
  timestampFromFuture,
} from "./utils/helpers";
import {
  ANCHOR_DISCRIMINATOR_LEN,
  DUMMY_ANCHOR_DISCRIMINATOR,
  TARGET_PROGRAM_PLACEHOLDER,
} from "./utils/constants";

describe("07 - Negative authorization", () => {
  const provider = anchor.AnchorProvider.env();
  anchor.setProvider(provider);
  const baseWallet = provider.wallet as anchor.Wallet;
  const program = anchor.workspace.r3SessionKeys as Program<R3SessionKeys>;

  before(async () => {
    await ensureProgramInitialized(program, baseWallet);
  });

  it("Rejects approve by a non-owner", async () => {
    const smartWalletOwner = Keypair.generate();
    const sessionKey = Keypair.generate();
    const userSmartWallet = sessionKeys.deriveUserSmartWalletPda(
      program.programId,
      smartWalletOwner.publicKey
    );

    await program.methods
      .createSmartWallet(smartWalletOwner.publicKey)
      .accounts({ admin: baseWallet.publicKey })
      .signers([baseWallet.payer])
      .rpc();

    await program.methods
      .createSession(
        sessionKey.publicKey,
        TARGET_PROGRAM_PLACEHOLDER,
        await timestampFromFuture(provider.connection),
        DUMMY_ANCHOR_DISCRIMINATOR,
        ANCHOR_DISCRIMINATOR_LEN
      )
      .accounts({
        sessionExecutor: baseWallet.publicKey,
        userSmartWallet,
      })
      .signers([baseWallet.payer])
      .rpc();

    const session = sessionKeys.deriveSessionPda(
      program.programId,
      userSmartWallet,
      sessionKey.publicKey
    );
    const wrongOwner = Keypair.generate();
    const error = await sendExpectError(
      program.methods
        .approveSession(sessionKey.publicKey)
        // Override auto-resolved PDAs to deliberately pair the wrong signer with this wallet.
        .accountsStrict({
          smartWalletOwner: wrongOwner.publicKey,
          programConfig: sessionKeys.deriveProgramConfigPda(program.programId),
          userSmartWallet,
          session,
        })
        .signers([baseWallet.payer, wrongOwner])
        .rpc()
    );

    assert.ok(error.includes("ConstraintSeeds"), error);
    const state = await program.account.session.fetch(session);
    assert.ok("waitingForApproval" in state.status);
  });

  it("Rejects revoke by a non-owner", async () => {
    const smartWalletOwner = Keypair.generate();
    const sessionKey = Keypair.generate();
    const userSmartWallet = sessionKeys.deriveUserSmartWalletPda(
      program.programId,
      smartWalletOwner.publicKey
    );

    await program.methods
      .createSmartWallet(smartWalletOwner.publicKey)
      .accounts({ admin: baseWallet.publicKey })
      .signers([baseWallet.payer])
      .rpc();

    await program.methods
      .createSession(
        sessionKey.publicKey,
        TARGET_PROGRAM_PLACEHOLDER,
        await timestampFromFuture(provider.connection),
        DUMMY_ANCHOR_DISCRIMINATOR,
        ANCHOR_DISCRIMINATOR_LEN
      )
      .accounts({
        sessionExecutor: baseWallet.publicKey,
        userSmartWallet,
      })
      .signers([baseWallet.payer])
      .rpc();

    const session = sessionKeys.deriveSessionPda(
      program.programId,
      userSmartWallet,
      sessionKey.publicKey
    );
    const wrongOwner = Keypair.generate();
    const error = await sendExpectError(
      program.methods
        .revokeSession(sessionKey.publicKey)
        // Override auto-resolved PDAs to deliberately pair the wrong signer with this wallet.
        .accountsStrict({
          smartWalletOwner: wrongOwner.publicKey,
          programConfig: sessionKeys.deriveProgramConfigPda(program.programId),
          userSmartWallet,
          session,
        })
        .signers([baseWallet.payer, wrongOwner])
        .rpc()
    );

    assert.ok(error.includes("ConstraintSeeds"), error);
    const state = await program.account.session.fetch(session);
    assert.ok("waitingForApproval" in state.status);
  });

  it("Rejects an executor the session was not created for", async () => {
    const smartWalletOwner = Keypair.generate();
    const sessionKey = Keypair.generate();
    const userSmartWallet = sessionKeys.deriveUserSmartWalletPda(
      program.programId,
      smartWalletOwner.publicKey
    );

    await program.methods
      .createSmartWallet(smartWalletOwner.publicKey)
      .accounts({ admin: baseWallet.publicKey })
      .signers([baseWallet.payer])
      .rpc();

    await program.methods
      .createSession(
        sessionKey.publicKey,
        TARGET_PROGRAM_PLACEHOLDER,
        await timestampFromFuture(provider.connection),
        DUMMY_ANCHOR_DISCRIMINATOR,
        ANCHOR_DISCRIMINATOR_LEN
      )
      .accounts({
        sessionExecutor: baseWallet.publicKey,
        userSmartWallet,
      })
      .signers([baseWallet.payer])
      .rpc();

    await program.methods
      .approveSession(sessionKey.publicKey)
      .accounts({ smartWalletOwner: smartWalletOwner.publicKey })
      .signers([baseWallet.payer, smartWalletOwner])
      .rpc();

    const wrongExecutor = Keypair.generate();
    const targetInstruction = new TransactionInstruction({
      programId: TARGET_PROGRAM_PLACEHOLDER,
      keys: [{ pubkey: userSmartWallet, isSigner: false, isWritable: false }],
      data: DUMMY_ANCHOR_DISCRIMINATOR,
    });
    const error = await sendExpectError(
      executeWithSession(program, {
        sessionExecutor: wrongExecutor.publicKey,
        sessionKey: sessionKey.publicKey,
        userSmartWallet,
        targetInstruction,
      })
        .signers([baseWallet.payer, wrongExecutor, sessionKey])
        .rpc()
    );

    assert.ok(error.includes("UnauthorizedSessionExecutor"), error);
  });

  it("Rejects pause by a non-admin", async () => {
    const wrongAdmin = Keypair.generate();
    const error = await sendExpectError(
      program.methods
        .pause()
        .accounts({ admin: wrongAdmin.publicKey })
        .signers([baseWallet.payer, wrongAdmin])
        .rpc()
    );

    assert.ok(error.includes("UnauthorizedAdmin"), error);
    const config = await program.account.programConfig.fetch(
      sessionKeys.deriveProgramConfigPda(program.programId)
    );
    assert.ok("active" in config.status);
  });

  it("Rejects unpause by a non-admin", async () => {
    await program.methods
      .pause()
      .accounts({ admin: baseWallet.publicKey })
      .signers([baseWallet.payer])
      .rpc();

    try {
      const wrongAdmin = Keypair.generate();
      const error = await sendExpectError(
        program.methods
          .unpause()
          .accounts({ admin: wrongAdmin.publicKey })
          .signers([baseWallet.payer, wrongAdmin])
          .rpc()
      );

      assert.ok(error.includes("UnauthorizedAdmin"), error);
      const config = await program.account.programConfig.fetch(
        sessionKeys.deriveProgramConfigPda(program.programId)
      );
      assert.ok("paused" in config.status);
    } finally {
      await program.methods
        .unpause()
        .accounts({ admin: baseWallet.publicKey })
        .signers([baseWallet.payer])
        .rpc();
    }
  });
});

File: programs/r3-session-keys/tests/07-negative_authorization.rs
mod common;

use {
    anchor_lang::{prelude::AccountMeta, solana_program::instruction::Instruction},
    common::{
        client, load, send_tx_expect_error, send_tx_expect_ok, timestamp_from_future, Env,
        DUMMY_ANCHOR_DISCRIMINATOR, TARGET_PROGRAM_PLACEHOLDER,
    },
    r3_session_keys::state::{
        program_config::{ProgramConfig, ProgramStatus},
        session::{Session, SessionStatus},
    },
    solana_keypair::Keypair,
    solana_signer::Signer,
};

struct SessionFixture {
    smart_wallet_owner: Keypair,
    session_key: Keypair,
    user_smart_wallet: anchor_lang::prelude::Pubkey,
    session: anchor_lang::prelude::Pubkey,
}

fn initialize(env: &mut Env, client: &client::R3SessionKeysClient) {
    let ix = client.initialize(env.admin.pubkey());
    send_tx_expect_ok(&mut env.svm, ix, &[&env.admin]);
}

fn create_session_fixture(env: &mut Env, client: &client::R3SessionKeysClient) -> SessionFixture {
    let smart_wallet_owner = Keypair::new();
    let session_key = Keypair::new();

    let (ix, user_smart_wallet, _) =
        client.create_smart_wallet(env.admin.pubkey(), smart_wallet_owner.pubkey());
    send_tx_expect_ok(&mut env.svm, ix, &[&env.admin]);

    let (ix, session, _) = client.create_session(
        env.admin.pubkey(),
        user_smart_wallet,
        session_key.pubkey(),
        TARGET_PROGRAM_PLACEHOLDER,
        timestamp_from_future(),
        DUMMY_ANCHOR_DISCRIMINATOR.to_vec(),
        DUMMY_ANCHOR_DISCRIMINATOR.len() as u8,
    );
    send_tx_expect_ok(&mut env.svm, ix, &[&env.admin]);

    SessionFixture {
        smart_wallet_owner,
        session_key,
        user_smart_wallet,
        session,
    }
}

fn assert_error(error: &str, expected_code: &str) {
    assert!(
        error.contains(&format!("Error Code: {expected_code}")),
        "expected {expected_code}, got:\n{error}"
    );
}

#[test]
fn test_approve_rejects_non_smart_wallet_owner() {
    let client = client::R3SessionKeysClient::new();
    let mut env = Env::new();
    initialize(&mut env, &client);
    let fixture = create_session_fixture(&mut env, &client);
    let wrong_owner = Keypair::new();

    let ix = client.approve_session(
        wrong_owner.pubkey(),
        fixture.user_smart_wallet,
        fixture.session_key.pubkey(),
    );
    let error = send_tx_expect_error(&mut env.svm, ix, &[&env.admin, &wrong_owner]);

    assert_error(&error, "ConstraintSeeds");
    let session: Session = load(&env, &fixture.session);
    assert_eq!(session.status, SessionStatus::WaitingForApproval);
}

#[test]
fn test_revoke_rejects_non_smart_wallet_owner() {
    let client = client::R3SessionKeysClient::new();
    let mut env = Env::new();
    initialize(&mut env, &client);
    let fixture = create_session_fixture(&mut env, &client);
    let wrong_owner = Keypair::new();

    let ix = client.revoke_session(
        wrong_owner.pubkey(),
        fixture.user_smart_wallet,
        fixture.session_key.pubkey(),
    );
    let error = send_tx_expect_error(&mut env.svm, ix, &[&env.admin, &wrong_owner]);

    assert_error(&error, "ConstraintSeeds");
    let session: Session = load(&env, &fixture.session);
    assert_eq!(session.status, SessionStatus::WaitingForApproval);
}

#[test]
fn test_execute_rejects_unauthorized_session_executor() {
    let client = client::R3SessionKeysClient::new();
    let mut env = Env::new();
    initialize(&mut env, &client);
    let fixture = create_session_fixture(&mut env, &client);

    let approve = client.approve_session(
        fixture.smart_wallet_owner.pubkey(),
        fixture.user_smart_wallet,
        fixture.session_key.pubkey(),
    );
    send_tx_expect_ok(
        &mut env.svm,
        approve,
        &[&env.admin, &fixture.smart_wallet_owner],
    );

    let wrong_executor = Keypair::new();
    let target_ix = Instruction {
        program_id: TARGET_PROGRAM_PLACEHOLDER,
        accounts: vec![AccountMeta::new_readonly(fixture.user_smart_wallet, false)],
        data: DUMMY_ANCHOR_DISCRIMINATOR.to_vec(),
    };
    let execute = client.execute_with_session(
        wrong_executor.pubkey(),
        fixture.session_key.pubkey(),
        fixture.user_smart_wallet,
        target_ix,
    );
    let error = send_tx_expect_error(
        &mut env.svm,
        execute,
        &[&env.admin, &wrong_executor, &fixture.session_key],
    );

    assert_error(&error, "UnauthorizedSessionExecutor");
}

#[test]
fn test_pause_rejects_non_admin() {
    let client = client::R3SessionKeysClient::new();
    let mut env = Env::new();
    initialize(&mut env, &client);
    let wrong_admin = Keypair::new();

    let ix = client.pause(wrong_admin.pubkey());
    let error = send_tx_expect_error(&mut env.svm, ix, &[&env.admin, &wrong_admin]);

    assert_error(&error, "UnauthorizedAdmin");
    let config: ProgramConfig = load(&env, &env.program_config);
    assert_eq!(config.status, ProgramStatus::Active);
}

#[test]
fn test_unpause_rejects_non_admin() {
    let client = client::R3SessionKeysClient::new();
    let mut env = Env::new();
    initialize(&mut env, &client);

    let pause = client.pause(env.admin.pubkey());
    send_tx_expect_ok(&mut env.svm, pause, &[&env.admin]);

    let wrong_admin = Keypair::new();
    let unpause = client.unpause(wrong_admin.pubkey());
    let error = send_tx_expect_error(&mut env.svm, unpause, &[&env.admin, &wrong_admin]);

    assert_error(&error, "UnauthorizedAdmin");
    let config: ProgramConfig = load(&env, &env.program_config);
    assert_eq!(config.status, ProgramStatus::Paused);
}
